packages/melody-streams/src/component.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2018 trivago N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { enqueueComponent } from 'melody-idom';
+import { shallowEqual } from './util/shallowEqual';
+import { BehaviorSubject, Subject } from 'rxjs';
+import { distinctUntilChanged } from 'rxjs/operators';
+
+function Component(element) {
+    // part of the public API
+    this.el = element;
+    this.refs = {};
+    this.props = new BehaviorSubject({});
+    this.updates = new Subject();
+    this.subscriptions = [];
+    this.state = {};
+}
+
+Object.assign(Component.prototype, {
+    apply(props) {
+        this.props.next(props);
+        if (this.subscriptions.length === 0) {
+            const t = this.getTransform({
+                dispatch(eventName, detail, options = {}) {
+                    const event = new CustomEvent(eventName, {
+                        ...options,
+                        detail,
+                    });
+                    this.el.dispatchEvent(event);
+                },
+                props: this.props,
+                updates: this.updates,
+                subscribe: obs => this.subscriptions.push(obs.subscribe()),
+            });
+            const s = t
+                .pipe(distinctUntilChanged(shallowEqual))
+                .subscribe(state => {
```

Comment:
1. what happens when an error occurs?
2. what happens when the stream completes early?
3. would it make sense to check for a valid state shape (plain object with only primitives as values) here?